### PR TITLE
Start documenting the Objective-C ABI

### PR DESCRIPTION
**Pull request by jacob-carlborg**
This does not include the complete ABI, only what's necessary to call instance methods and class methods. This document will be extended when additional features are added.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
@JinShil @ibuclaw @WalterBright @wilzbach @UplinkCoder please have a look.

File: docs/objective-c_abi.md [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+language constructs.
+
+## Messages
+
+The Objective-C model of object-oriented programming is based on message passing
+to object instances. Unlike D or C++ where a method is called. The difference
+from implementation stand point is that in D and C++ a vtable is used which
+is an array of function pointers and the compiler uses an index into that array
+to determine which method to call. In Objective-C it's the runtime that is
+responsible to find the correct implementation when a message is sent to an
```
Comment (JinShil) [inline docs/objective-c_abi.md]: s/to find/for finding

File: docs/objective-c_abi.md [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+* The last parameter is for all the arguments that the
+  implementation expects
+
+The above example would be translated by the compiler to the following:
+
+```c
+int result = objc_msgSend(receiver, selector);
+```
+
+If the method expects any arguments they're passed after the selector argument:
```
Comment (JinShil) [inline docs/objective-c_abi.md]: s/they're/it's

File: docs/objective-c_abi.md [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+The compiler implements this by making a regular C call to the `objc_msgSend`
+function in the Objective-C runtime. The signature of `objc_msgSend` looks
+something like this:
+
+```objective-c
+id objc_msgSend(id self, SEL op, ...);
+```
+
+* The first parameter is the receiver (this/self pointer)
```
Comment (JinShil) [inline docs/objective-c_abi.md]: should "this" and "self" be surrounded by tick marks?

File: docs/objective-c_abi.md [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+If the method expects any arguments they're passed after the selector argument:
+
+```c
+int result = objc_msgSend(receiver, selector, arg1, arg2);
+```
+
+The call to `objc_msgSend` should not be performed as a variadic call but
+instead as if it had the same signature as the method that should be called but
+with the two additional parameter, `self` and `op`, added first. The
```
Comment (JinShil) [inline docs/objective-c_abi.md]: That first sentence is a little hard to follow.  Maybe it can be rephrased.

File: docs/objective-c_abi.md [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+implementation of `objc_msgSend` will jump to the method instead of calling it.
+
+Because of the above, multiple versions of `objc_msgSend` exist. Depending on
+the return type of the method that is called the correct version need to be
```
Comment (JinShil) [inline docs/objective-c_abi.md]: comma after "called". 

s/the correct version need to be used/the correct version will be used

File: docs/objective-c_abi.md [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+used. This depends on the platform C ABI. This is a list of functions and for
+which return types they're used:
```
Comment (JinShil) [inline docs/objective-c_abi.md]: remove "and"?

File: docs/objective-c_abi.md [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+registries
+* `objc_msgSend_fpret` - Used for `long double`
+* `objc_msgSend_fp2ret` - Used for `_Complex long double`
+* `objc_msgSend` - Used for everything else
+
+### Returning a Struct
+
+If a struct is small enough to be returned in registers (according to the
+platform C ABI) the regular `objc_msgSend` function is used. If the struct will
+not fit in registers the `objc_msgSend_stret` function is used. The signature of
```
Comment (JinShil) [inline docs/objective-c_abi.md]: Could use a comma after the "if" clauses

File: docs/objective-c_abi.md [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+
+### Metaclass
+
+All classes in Objective-C are themselves objects. A class object is an
+instance of the class' metaclass. Metaclasses follow their own inheritance
+chain. A metaclass inherits from the metaclass of the class' superclass. This
+continues all the way up to the root class (which in most cases is the NSObject
+class). The metaclass of the root class is an instance of itself.
+
+Below follows a diagram of the inheritance and instance relationships between
```
Comment (JinShil) [inline docs/objective-c_abi.md]: s/follows/is

File: docs/objective-c_abi.md [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+
+#### Private Linkage
+
+Like [Internal](#internal-linkage), but omit from symbol table.
+
+### `L_OBJC_METH_VAR_NAME_`
+
+For each selector that is used, a symbol is generated in the resulting binary.
+The symbol has the name `L_OBJC_METH_VAR_NAME_.<number>`, where `<number>` is an
+incrementing number. The selector is stored as a null terminated C string as the
```
Comment (JinShil) [inline docs/objective-c_abi.md]: s/as the section data/in the data section?

File: docs/objective-c_abi.md [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+| Section                                    | Linkage                      | Alignment |
+|--------------------------------------------|------------------------------|-----------|
+| [`__objc_selrefs`](#segments-and-sections) | [Private](#private-linkage)  | 8         |
+
+### `L_OBJC_CLASSLIST_REFERENCES_$_`
+
+For each externally defined class that is referenced, a symbol is generated in
+the resulting binary. The symbol has the name
+`L_OBJC_CLASSLIST_REFERENCES_$_.<number>`, where `<number>` is an incrementing
+number. The content of the symbols is a reference to an externally defined
```
Comment (JinShil) [inline docs/objective-c_abi.md]: Should "symbols" be singular here?

File: docs/objective-c_abi.md [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+symbol with the name `_OBJC_CLASS_$_<class_name>`, where `<class_name>` is the
+name of the class.
+
+| Section                                      | Linkage                      | Alignment |
+|----------------------------------------------|------------------------------|-----------|
+| [`__objc_classrefs`](#segments-and-sections) | [Private](#private-linkage)  | 8         |
+
+### `L_OBJC_IMAGE_INFO`
+
+For any binary that is built the `L_OBJC_IMAGE_INFO` symbols is generated. The
```
Comment (JinShil) [inline docs/objective-c_abi.md]: comma after built?

File: docs/objective-c_abi.md [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+
+#### `version`
+
+Seems to always be fixed.
+
+#### `flags`
+
+Indicates if features like: garbage collector, automatic reference counting
+(ARC) or class properties are supported. These features can be enabled/disabled
+in the Clang compiler using command line switches. The exact values used or the
```
Comment (JinShil) [inline docs/objective-c_abi.md]: comma after "used" and "supported"?

File: docs/objective-c_abi.md [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+| `__objc_classlist` | `__DATA`  | `regular`           | `no_dead_strip` | 8          |
+| `__objc_selrefs`   | `__DATA`  | `literal_pointers`  | `no_dead_strip` | 8          |
+| `__objc_classrefs` | `__DATA`  | `regular`           | `no_dead_strip` | 8          |
+
+For more information about the different section types and attributes, see
+the documentation for [Assembler Directives](https://developer.apple.com/library/content/documentation/DeveloperTools/Reference/Assembler/040-Assembler_Directives/asm_directives.html) from Apple.
+
+## Tools
+
+Here follows a list of useful tools that can/has been used to get the
```
Comment (JinShil) [inline docs/objective-c_abi.md]: s/has/have

**Review by JinShil — APPROVED** [review]
I can't say I understand all of this, but if I were maintaining the Objective-C features of DMD, I think it would come in handy.  My review is mostly just a bunch of nits.

File: docs/objective-c_abi.md (hunk 2 above)
Comment (jacob-carlborg) [reply docs/objective-c_abi.md]: But `arguments` is plural?

Comment (JinShil) [reply docs/objective-c_abi.md]: Sorry, I don't know what I was thinking.  Just ignore anything stupid in my comments.

File: docs/objective-c_abi.md (hunk 4 above)
Comment (jacob-carlborg) [reply docs/objective-c_abi.md]: I changed it slightly, but I'm not sure if it's any better. Writing is not my strongest skill. Suggestions are welcome.

**Comment by jacob-carlborg** [comment]
Fixed all comments, except two, where I asked for more input.

**Review by JinShil — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
@JinShil @wilzbach can we please merge this as well.
